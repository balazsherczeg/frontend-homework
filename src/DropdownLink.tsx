import { connect } from 'react-redux'
import { createStructuredSelector } from 'reselect'

import { ReduxState, Company } from './types'
import { getIsDropdownMenuVisible, getSelectedCompany } from './selectors'
import { toggleDropdownMenuVisibility } from './actions'

import OutsideClicker from './OutsideClicker'
import DropdownMenu from './DropdownMenu'
import Animator from './Animator'

type ReduxProps = {
  isDropdownMenuVisible: boolean,
  selectedCompany?: Company,
}

type DispatchProps = {
  toggleDropdownMenuVisibility: () => void,
}

export const DropdownLink = ({
  isDropdownMenuVisible,
  toggleDropdownMenuVisibility,
  selectedCompany,
}: ReduxProps & DispatchProps) => (
  <div className="DropdownLink__wrapper">
    <div className="DropdownLink" onClick={toggleDropdownMenuVisibility} data-test-nav-link>
      <div className="DropdownLink__text-wrapper">
        <div className="DropdownLink__text">
          Elon Musk
        </div>

        {selectedCompany && (
          <div className="DropdownLink__subtext">
            {selectedCompany.name}
          </div>
        )}
      </div>

      <i className="material-icons-outlined DropdownLink__icon">
        settings
      </i>
    </div>

    {isDropdownMenuVisible && <OutsideClicker />}

    <Animator
      className="DropdownAnimator"
      isMounted={isDropdownMenuVisible}
    >
      <DropdownMenu />
    </Animator>
  </div>
)

export default connect(
  createStructuredSelector<ReduxState, ReduxProps>({
    isDropdownMenuVisible: getIsDropdownMenuVisible,
    selectedCompany: getSelectedCompany,
  }),
  { toggleDropdownMenuVisibility }
)(DropdownLink)
