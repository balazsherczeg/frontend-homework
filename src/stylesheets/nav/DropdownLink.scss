.DropdownLink__wrapper {
  position: relative;
}

.DropdownLink {
  display: flex;
  align-items: center;
  max-width: 280px;
  padding: 8px;

  border-radius: 8px;

  cursor: pointer;

  transition: 100ms background-color;

  &:hover, &:focus {
    background-color: $grey-2;
  }

  &:active {
    background-color: $grey-3;
  }
}

.DropdownLink__text-wrapper {
  text-align: right;
}

.DropdownLink__text {
  font-weight: bold;
}

.DropdownLink__subtext {
  color: $grey-6;
}

.DropdownLink__icon {
  margin-left: 8px;
}

