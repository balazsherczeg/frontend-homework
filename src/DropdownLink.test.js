import { shallow } from 'enzyme'
import { DropdownLink } from './DropdownLink'

const selectedCompany = { name: 'Viljatootja AS' }

describe('<DropdownLink />', () => {
  const toggleDropdownMenuVisibilityMock = jest.fn()

  const render = (props) => shallow(
    <DropdownLink
      toggleDropdownMenuVisibility={toggleDropdownMenuVisibilityMock}
      {...props}
    />
  )

  it('renders without dropdown menu', () => {
    expect(render({ isDropdownMenuVisible: false, selectedCompany })).toMatchSnapshot()
  })

  it('renders with dropdown menu', () => {
    expect(render({ isDropdownMenuVisible: true, selectedCompany })).toMatchSnapshot()
  })

  it('calls toggleDropdownMenuVisibility when nav link is clicked', () => {
    const component = render({ isDropdownMenuVisible: false, selectedCompany })
    component.find('[data-test-nav-link]').simulate('click')

    expect(toggleDropdownMenuVisibilityMock).toHaveBeenCalled()
  })
})
