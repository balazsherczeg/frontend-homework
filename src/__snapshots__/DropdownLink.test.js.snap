// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownLink /> renders with dropdown menu 1`] = `
<div
  className="DropdownLink__wrapper"
>
  <div
    className="DropdownLink"
    data-test-nav-link={true}
    onClick={[MockFunction]}
  >
    <div
      className="DropdownLink__text-wrapper"
    >
      <div
        className="DropdownLink__text"
      >
        Elon Musk
      </div>
      <div
        className="DropdownLink__subtext"
      >
        Viljatootja AS
      </div>
    </div>
    <i
      className="material-icons-outlined DropdownLink__icon"
    >
      settings
    </i>
  </div>
  <OutsideClicker />
  <Animator
    className="DropdownAnimator"
    isMounted={true}
  >
    <DropdownMenu />
  </Animator>
</div>
`;

exports[`<DropdownLink /> renders without dropdown menu 1`] = `
<div
  className="DropdownLink__wrapper"
>
  <div
    className="DropdownLink"
    data-test-nav-link={true}
    onClick={[MockFunction]}
  >
    <div
      className="DropdownLink__text-wrapper"
    >
      <div
        className="DropdownLink__text"
      >
        Elon Musk
      </div>
      <div
        className="DropdownLink__subtext"
      >
        Viljatootja AS
      </div>
    </div>
    <i
      className="material-icons-outlined DropdownLink__icon"
    >
      settings
    </i>
  </div>
  <Animator
    className="DropdownAnimator"
    isMounted={false}
  >
    <DropdownMenu />
  </Animator>
</div>
`;
